File: 0189-rotate-array/0189-rotate-array.java
class Solution {
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k %= nums.length;
        rev(nums, 0, n-1);
        rev(nums, 0, k-1);
        rev(nums, k, n-1);
    }
    
    public static void rev(int[] nums, int start, int end) {
        while(start < end) {
            int temp = nums[start];
            nums[start] = nums[end];
            nums[end] = temp;
            start++;
            end--;
        }
    }
}

File: 0189-rotate-array/0189-rotate-array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int n = nums.size();
        // if(n == 1) return;
        if(k >= n) k = k % n;
        vector<int> temp;
        
        // storing the last k variable into temp
        for(int i = n-k; i < n; i++) {
            temp.push_back(nums[i]);
        }
        
        // shifting
        for(int i = n-1; i >= k; i--) {
            nums[i] = nums[i-k];            
        }
        
        // put back temp
        int j = 0;
        for(int i = 0; i < k; i++) {
            nums[i] = temp[j];
            j++;
        }
    }
};